File: screenit-se/se-proxy/src/main/java/nl/rivm/screenit/mamma/se/proxy/util/DateUtil.java
package nl.rivm.screenit.mamma.se.proxy.util;

/*-
 * ========================LICENSE_START=================================
 * screenit-se-proxy
 * %%
 * Copyright (C) 2012 - 2025 Facilitaire Samenwerking Bevolkingsonderzoek
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * =========================LICENSE_END==================================
 */

import java.time.Duration;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Date;

import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class DateUtil
{
	public static final ZoneId SCREENIT_DEFAULT_ZONE = ZoneId.of("Europe/Amsterdam");

	private static Duration offset = Duration.ZERO;

	private static final DateTimeFormatter DEFAULT_DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("dd-MM-yyyy HH:mm:ss");

	public static void setOffset(Duration newOffset)
	{
		LOG.info("De tijd is gewijzigd van {} naar {}", getCurrentDateTime(), ZonedDateTime.now(SCREENIT_DEFAULT_ZONE).plus(newOffset));
		DateUtil.offset = newOffset;
	}

	public static LocalDateTime getCurrentDateTime()
	{
		return LocalDateTime.now(SCREENIT_DEFAULT_ZONE).plus(DateUtil.offset);
	}

	public static LocalDate getCurrentDate()
	{
		return getCurrentDateTime().toLocalDate();
	}

	public static boolean isVandaag(LocalDate localDate)
	{
		return getCurrentDateTime().toLocalDate().isEqual(localDate);
	}

	public static LocalDateTime getAfgelopenMiddernacht()
	{
		return getCurrentDateTime().toLocalDate().atStartOfDay();
	}

	public static Duration getOffset()
	{
		return DateUtil.offset;
	}

	public static String format(LocalDateTime localDateTime)
	{
		return DEFAULT_DATE_TIME_FORMATTER.format(localDateTime);
	}

	public static Date toUtilDate(LocalDate localDate)
	{
		return Date.from(localDate.atStartOfDay(SCREENIT_DEFAULT_ZONE).toInstant());
	}

	public static Date toUtilDate(LocalDateTime localDateTime)
	{
		return Date.from(localDateTime.atZone(SCREENIT_DEFAULT_ZONE).toInstant());
	}
}

File: screenit-se/se-proxy/webapp/src/util/DateUtil.ts
/*-
 * ========================LICENSE_START=================================
 * screenit-se-proxy
 * %%
 * Copyright (C) 2012 - 2025 Facilitaire Samenwerking Bevolkingsonderzoek
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * =========================LICENSE_END==================================
 */
import {createActionKiesDaglijstDatum} from "../actions/DaglijstDatumActions"
import {store} from "../Store"
import {leesAfspraken} from "../restclient/DaglijstRestclient"
import {leesPlanning} from "../restclient/PlanningRestClient"
import {createActionNavigateToDaglijst} from "../actions/NavigationActions"
import {add, differenceInMinutes, Duration, format, formatDistanceToNow, formatDuration, isAfter, isBefore, isEqual, sub} from "date-fns"

export const DATUM_FORMAT = "yyyy-MM-dd"
export const NL_DATUM_FORMAT = "dd-MM-yyyy"
export const TIJD_FORMAT = "HH:mm"
export const TIMESTAMP_FORMAT = "HH:mm:ss.SSS"
export const ISO_TIMESTAMP_FORMAT = "yyyy-MM-dd'T'HH:mm:ss"
export const NL_DATUM_TIJD_FORMAT = "dd-MM-yyyy HH:mm"

export const vandaagISO = (): string => format(nu(), DATUM_FORMAT)
let offset: Duration | null = null

export const setOffset = (val: Duration): void => {
	offset = val

	if (store.getState().session) {
		store.dispatch(createActionKiesDaglijstDatum(vandaagISO()))
		leesAfspraken(vandaagISO(), createActionNavigateToDaglijst())
		leesPlanning(vandaagISO())
	}
}

export const nu = (): Date => {
	if (offset !== null) {
		return add(new Date(), offset)
	}

	return new Date()
}

export const nuISO = (): string => format(nu(), ISO_TIMESTAMP_FORMAT)

export const nuTimestamp = (): string => format(nu(), TIMESTAMP_FORMAT)

export const nuTijdUrenMinuten = (): string => format(nu(), TIJD_FORMAT)

export const ligtTussenData = (
	datum: Date,
	startDatum: Date | null,
	eindDatum: Date | null,
): boolean => {
	if (startDatum && eindDatum) {
		return (
			(isAfter(datum, startDatum) || isEqual(datum, startDatum)) &&
			(isBefore(datum, eindDatum) || isEqual(datum, eindDatum))
		)
	}
	if (startDatum) {
		return isAfter(datum, startDatum) || isEqual(datum, startDatum)
	}
	if (eindDatum) {
		return isBefore(datum, eindDatum) || isEqual(datum, eindDatum)
	}
	return false
}

export const datumFormaat = (isoDatum: string | Date | null | undefined): string => isoDatum ? format(new Date(String(isoDatum)), NL_DATUM_FORMAT) : ""

export const tijdFormaat = (isoTijd: string | undefined): string => isoTijd ? format(new Date(isoTijd), TIJD_FORMAT) : ""

export const datumTijdFormaat = (isoTijd: string | undefined): string => isoTijd ? format(new Date(isoTijd), NL_DATUM_TIJD_FORMAT) : ""

export const getDate = (isoDatetime: string): string => {
	return isoDatetime.split("T")[0]
}

export const getTime = (isoDatetime: string): string => {
	return isoDatetime.split("T")[1].slice(0, 5)
}

export const datumInVerleden = (isoDateTime: string): boolean => {
	const datum = new Date(isoDateTime)
	const gisteren = sub(nu(), {days: 1})
	return datum <= gisteren
}

export const datumInToekomst = (isoDateTime: string): boolean => {
	const datum = new Date(isoDateTime)
	const vandaagIso = nu()
	return datum > vandaagIso
}

export const getTijdGeledenTekst = (timestamp: string): string => {
	return formatDistanceToNow(new Date(timestamp), {addSuffix: true})
}

export const vandaagPlusDagen = (aantal: number): Date => {
	return add(nu(), {days: aantal})
}

export const vandaagDate = (): Date => {
	return new Date(vandaagISO())
}

export const isValideDatum = (datum: Date): boolean => {
	return datum.getFullYear().toString().length === 4
}

export const stringDatumEnStringTijdNaarDatumTijd = (datum: string, tijd: string): Date => {
	return new Date(`${datum}T${tijd}`)
}

export const verschilDatumTijdEnNuInMinuten = (datumTijd: string | Date): number => {
	return differenceInMinutes(new Date(nuISO()), new Date(datumTijd))
}

export const numberMinutenNaarStringUrenEnMinuten = (minuten: number): string => {
	return formatDuration({minutes: minuten})
}

export const javaDurationToJavascript = (durationStr: string): Duration => {
	const match = durationStr.match(
		/^P(?:(-?\d+)W)?(?:(-?\d+)D)?(?:T(?:(-?\d+)H)?(?:(-?\d+)M)?(?:(-?\d+(?:\.\d+)?)S)?)?$/,
	)
	if (!match) {
		throw new Error(`Invalid Java Duration format: ${durationStr}`)
	}
	const [, w, d, h, m, s] = match
	const days = (w ? parseInt(w, 10) * 7 : 0) + (d ? parseInt(d, 10) : 0)
	return {
		days,
		hours: h ? parseInt(h, 10) : 0,
		minutes: m ? parseInt(m, 10) : 0,
		seconds: s ? parseFloat(s) : 0,
	}
}
